File: androidApp/src/main/kotlin/com/humblesolutions/aromex/data/BackendCommissionRuleRepository.kt
package com.humblesolutions.aromex.data

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.SetOptions
import com.google.firebase.ktx.Firebase
import com.google.firebase.firestore.ktx.firestore
import com.humblesolutions.aromex.model.CommissionRule
import com.humblesolutions.aromex.model.CommissionRuleInput
import com.humblesolutions.aromex.model.FirebaseClientConfig
import com.humblesolutions.aromex.model.RateKind
import com.humblesolutions.aromex.repository.CommissionRuleRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

/**
 * Firestore-backed [CommissionRuleRepository] — `commissionRules/{ruleId}` standing arrangements
 * (ticket #97). Firebase-only (a rule touches no Humble Ledger). Every write is reached only
 * through the shared commission-rule use cases, which gate on `role == ADMIN`; the mobile
 * Firestore rules are the backstop.
 */
class BackendCommissionRuleRepository(
    private val context: Context,
    private val config: FirebaseClientConfig,
) : CommissionRuleRepository {

    private val app get() = FirebaseAppFactory.get(context, config)
    private val db get() = Firebase.firestore(app)
    private fun currentUid(): String? = FirebaseAuth.getInstance(app).currentUser?.uid

    override fun observeRules(includeInactive: Boolean): Flow<List<CommissionRule>> = callbackFlow {
        val query = if (includeInactive) {
            db.collection(COLLECTION)
        } else {
            db.collection(COLLECTION).whereEqualTo("isActive", true)
        }
        val registration = query.addSnapshotListener { snapshot, error ->
            if (error != null) {
                close(error)
                return@addSnapshotListener
            }
            trySend(snapshot?.documents?.mapNotNull { it.toRule() }.orEmpty())
        }
        awaitClose { registration.remove() }
    }

    override suspend fun saveRule(input: CommissionRuleInput): String = withContext(Dispatchers.IO) {
        val editing = input.ruleId.isNotBlank()
        val ref = if (editing) db.collection(COLLECTION).document(input.ruleId)
        else db.collection(COLLECTION).document()
        // Editing keeps the same doc so earned commissions keep pointing at a live rule; a
        // create seeds createdBy/createdAt, an edit only bumps updatedAt.
        val base = mutableMapOf<String, Any?>(
            "ruleId" to ref.id,
            "locationAttributeId" to input.locationAttributeId,
            "payeeEntityId" to input.payeeEntityId,
            "rateKind" to input.rateKind.name,
            "rate" to input.rate, // decimal String — never a Double
            "isActive" to input.isActive,
            "updatedAt" to FieldValue.serverTimestamp(),
        )
        if (!editing) {
            base["createdBy"] = currentUid()
            base["createdAt"] = FieldValue.serverTimestamp()
        }
        ref.set(base, SetOptions.merge()).await()
        ref.id
    }

    override suspend fun archiveRule(ruleId: String): Unit = withContext(Dispatchers.IO) {
        db.collection(COLLECTION).document(ruleId)
            .update(mapOf("isActive" to false, "updatedAt" to FieldValue.serverTimestamp()))
            .await()
    }

    private fun DocumentSnapshot.toRule(): CommissionRule? {
        if (!exists()) return null
        val rateKind = RateKind.fromWire(getString("rateKind") ?: "") ?: return null
        return CommissionRule(
            ruleId = getString("ruleId") ?: id,
            locationAttributeId = getString("locationAttributeId").orEmpty(),
            payeeEntityId = getString("payeeEntityId").orEmpty(),
            rateKind = rateKind,
            rate = getString("rate") ?: "0",
            isActive = getBoolean("isActive") ?: true,
            createdBy = getString("createdBy"),
            createdAt = getTimestamp("createdAt")?.toDate()?.time,
            updatedAt = getTimestamp("updatedAt")?.toDate()?.time,
        )
    }

    private companion object {
        const val COLLECTION = "commissionRules"
    }
}

File: iosApp/iosApp/repository/BackendCommissionRuleRepository.swift
import Foundation
import FirebaseFirestore
import SharedLogic

/// Firestore-backed implementation of the shared `CommissionRuleRepository` (ticket #97).
///
/// ── Flow-interop (SKIE) ──────────────────────────────────────────────────────
/// `observeRules(includeInactive:)` returns a Kotlin `Flow<List<CommissionRule>>`, built with
/// the shared `commissionRulesCallbackFlow(subscribe:)` adapter feeding a Firestore snapshot
/// listener — so the shared use cases (admin gate for management, inventory gate for intake)
/// stay the single source of the list rules. SKIE `__`-prefixes the suspend members.
///
/// Firebase-only: a commission rule touches no Humble Ledger; it only decides what a later
/// intake will owe. Writes are reached only through the admin-gated shared use cases.
final class BackendCommissionRuleRepository: NSObject, CommissionRuleRepository {

    private let uid: String
    private let collection: CollectionReference

    init(config: FirebaseClientConfig, uid: String) {
        self.uid = uid
        self.collection = FirebaseAppFactory.firestore(for: config).collection("commissionRules")
        super.init()
    }

    // MARK: - Live read (produces the shared Flow the use case consumes)

    func observeRules(includeInactive: Bool) -> SkieSwiftFlow<[CommissionRule]> {
        return commissionRulesCallbackFlow { onEach, onError in
            let query: Query = includeInactive
                ? self.collection
                : self.collection.whereField("isActive", isEqualTo: true)
            let registration = query.addSnapshotListener { snapshot, error in
                if let error = error {
                    _ = onError(error.localizedDescription)
                    return
                }
                let rules = snapshot?.documents.compactMap {
                    Self.mapDoc(id: $0.documentID, data: $0.data())
                } ?? []
                _ = onEach(rules)
            }
            return FirestoreCommissionRuleObservation(registration)
        }
    }

    // MARK: - Suspend writes (consumed by the admin-gated shared use cases)

    func __saveRule(input: CommissionRuleInput) async throws -> String {
        let editing = !input.ruleId.isEmpty
        let ref = editing ? collection.document(input.ruleId) : collection.document()
        // Editing keeps the same doc so earned commissions keep pointing at a live rule; a
        // create seeds createdBy/createdAt, an edit only bumps updatedAt.
        var data: [String: Any] = [
            "ruleId": ref.documentID,
            "locationAttributeId": input.locationAttributeId,
            "payeeEntityId": input.payeeEntityId,
            "rateKind": input.rateKind.name,
            "rate": input.rate,            // decimal String — never a Double
            "isActive": input.isActive,
            "updatedAt": FieldValue.serverTimestamp(),
        ]
        if !editing {
            data["createdBy"] = uid
            data["createdAt"] = FieldValue.serverTimestamp()
        }
        try await ref.setData(data, merge: true)
        return ref.documentID
    }

    func __archiveRule(ruleId: String) async throws {
        try await collection.document(ruleId).updateData([
            "isActive": false,
            "updatedAt": FieldValue.serverTimestamp(),
        ])
    }

    func close() {}

    // MARK: - Mapping

    static func mapDoc(id: String, data: [String: Any]) -> CommissionRule? {
        guard let rateKind = RateKind.companion.fromWire(value: (data["rateKind"] as? String) ?? "") else {
            return nil
        }
        return CommissionRule(
            ruleId: (data["ruleId"] as? String) ?? id,
            locationAttributeId: (data["locationAttributeId"] as? String) ?? "",
            payeeEntityId: (data["payeeEntityId"] as? String) ?? "",
            rateKind: rateKind,
            rate: (data["rate"] as? String) ?? "0",
            isActive: (data["isActive"] as? Bool) ?? true,
            createdBy: data["createdBy"] as? String,
            createdAt: nil,
            updatedAt: nil
        )
    }
}

/// Detaches the Firestore snapshot listener when the shared observe `Flow` is cancelled.
private final class FirestoreCommissionRuleObservation: NSObject, CommissionRuleObservation {
    private let registration: ListenerRegistration
    init(_ registration: ListenerRegistration) { self.registration = registration }
    func cancel() { registration.remove() }
}
